src/Interpreters/ActionsDAG.cpp

```diff
@@ -3577,18 +3602,69 @@ static MutableColumnPtr deserializeConstant(
     return ColumnConst::create(std::move(column), 0);
 }
 
-void ActionsDAG::serialize(WriteBuffer & out, SerializedSetsRegistry & registry) const
+std::unordered_map<const ActionsDAG::Node *, size_t> ActionsDAG::getNodeToIdMap() const
 {
-    size_t nodes_size = nodes.size();
-    writeVarUInt(nodes_size, out);
-
     std::unordered_map<const Node *, size_t> node_to_id;
     for (const auto & node : nodes)
         node_to_id.emplace(&node, node_to_id.size());
 
-    if (nodes_size != node_to_id.size())
+    if (nodes.size() != node_to_id.size())
         throw Exception(ErrorCodes::LOGICAL_ERROR, "Duplicate nodes in ActionsDAG");
 
+    return node_to_id;
+}
+
+std::unordered_map<size_t, const ActionsDAG::Node *> ActionsDAG::getIdToNodeMap() const
```

Comment:
If to think about it, we can just return `std::vector<const ActionsDAG::Node *>`